components/tools/OmeroWeb/test/integration/test_api_projects.py

```diff
@@ -600,6 +600,87 @@ def test_project_update(self, user1):
         rsp = _get_response_json(django_client, project_url, {})
         assert rsp['data']['Description'] == 'New test description update'
 
+    def test_project_datasets_update(self, user1,
+                                     project_hierarchy_user1_group1):
+        """
+        Test updating a Project without losing child Datasets.
+
+        If we load a Project without loading Datasets, then update
+        and save the Project, we don't want to lose Dataset links
+        """
+        conn = get_connection(user1)
+        user_name = conn.getUser().getName()
+        django_client = self.new_django_client(user_name, user_name)
+
+        project = project_hierarchy_user1_group1[0]
+        dataset_count = len(project.linkedDatasetList())
+        dataset = project.linkedDatasetList()[0]
+        image_count = len(dataset.linkedImageList())
+
+        version = api_settings.API_VERSIONS[-1]
+        project_url = reverse('api_project',
+                              kwargs={'api_version': version,
+                                      'object_id': project.id.val})
+        save_url = reverse('api_save', kwargs={'api_version': version})
+        # Get Project, update and save back
+        rsp = _get_response_json(django_client, project_url, {})
+        project_json = rsp['data']
+        project_json['Name'] = 'renamed Project'
+        _csrf_put_json(django_client, save_url, project_json)
+
+        # Check Project has been updated and still has child Datasets
+        proj = conn.getObject('Project', project.id.val)
+        assert proj.getName() == 'renamed Project'
+        assert len(list(proj.listChildren())) == dataset_count
+
+        # Get Dataset, update and save back
+        dataset_url = reverse('api_dataset',
+                              kwargs={'api_version': version,
+                                      'object_id': dataset.id.val})
+        rsp = _get_response_json(django_client, dataset_url, {})
+        dataset_json = rsp['data']
+        dataset_json['Name'] = 'renamed Dataset'
+        _csrf_put_json(django_client, save_url, dataset_json)
+
+        # Check Dataset has been updated and still has child Images
+        dataset = conn.getObject('Dataset', dataset.id.val)
+        assert dataset.getName() == 'renamed Dataset'
+        assert len(list(dataset.listChildren())) == image_count
+
+    def test_project_tags_update(self, user1):
```

Comment:
I assumed that Datasets and Screens behave the same way as Projects when being linked to Tags etc. But if we need to test them all, I can parametrize that test.